tmol/kinematics/compiled/compiled.ops.cpp

```diff
@@ -0,0 +1,174 @@
+    saved_scans(scans, false),
+    saved_gens(gens, false),
+    saved_kintree(kintree, false)
+ { }
+
+  torch::autograd::variable_list apply(
+      torch::autograd::variable_list&& grads) override {
+    auto hts = saved_hts.unpack();
+    auto dofs = saved_dofs.unpack();
+    auto nodes = saved_nodes.unpack();
+    auto scans = saved_scans.unpack();
+    auto gens = saved_gens.unpack();
+    auto kintree = saved_kintree.unpack();
+
+    at::Tensor dV_ddof;
+    using Int = int32_t;
+    auto dVdx = grads[0];
+
+    TMOL_DISPATCH_FLOATING_DEVICE(
+      hts.type(), "kin_deriv_op", ([&] {
+        using Real = scalar_t;
+        constexpr tmol::Device Dev = device_t;
+
+        auto result = KinDerivDispatch<Dev, Real, Int>::f(
+            TCAST(dVdx),
+            TCAST(hts),
+            TCAST(dofs),
+            TCAST(nodes),
+            TCAST(scans),
+            TCAST(gens),
+            TCAST(kintree)
+        );
+
+        dV_ddof = result.tensor;
+      }));
+
+    return {dV_ddof};
+  }
+};
+
+
+Tensor kinematic_op(
+    Tensor dofs,
+    Tensor nodes_f,
+    Tensor scans_f,
+    Tensor gens_f,
+    Tensor nodes_b,
+    Tensor scans_b,
+    Tensor gens_b,
+    Tensor kintree
+) {
+  using tmol::utility::connect_backward_pass;
+
+  at::Tensor coords;
+  at::Tensor HTs;
+
+  using Int = int32_t;
+
+  TMOL_DISPATCH_FLOATING_DEVICE(
+      dofs.type(), "forward_kin_op", ([&] {
+        using Real = scalar_t;
+        constexpr tmol::Device Dev = device_t;
+
+        auto result = ForwardKinDispatch<Dev, Real, Int>::f(
+            TCAST(dofs),
+            TCAST(nodes_f),
+            TCAST(scans_f),
+            TCAST(gens_f),
+            TCAST(kintree));
+
+        coords = std::get<0>(result).tensor;
+        HTs = std::get<1>(result).tensor;
+      }));
+
+  return connect_backward_pass({dofs}, coords, [&]() {
+      return std::shared_ptr<KinOpBackward>(
+        new KinOpBackward( HTs, dofs, nodes_b, scans_b, gens_b, kintree ), 
+        torch::autograd::deleteFunction);
+  });
+};
+
+Tensor forward_only_op(
+    Tensor dofs,
+    Tensor nodes_f,
+    Tensor scans_f,
+    Tensor gens_f,
+    Tensor kintree
+) {
+  using tmol::utility::connect_backward_pass;
+
+  at::Tensor coords;
+  // at::Tensor HTs;
+
+  using Int = int32_t;
+
+  TMOL_DISPATCH_FLOATING_DEVICE(
+      dofs.type(), "forward_kin_only_op", ([&] {
+        using Real = scalar_t;
+        constexpr tmol::Device Dev = device_t;
+
+        auto result = ForwardKinDispatch<Dev, Real, Int>::f(
+            TCAST(dofs),
+            TCAST(nodes_f),
+            TCAST(scans_f),
+            TCAST(gens_f),
+            TCAST(kintree));
+
+        coords = std::get<0>(result).tensor;
+        // HTs = std::get<1>(result).tensor;
+      }));
+
+  return coords;
+
+};
+
+
+
+static auto registry =
+    torch::jit::RegisterOperators()
+        .op("tmol::forward_kin_op", &kinematic_op )
+        .op("tmol::forward_only_kin_op", &forward_only_op );
```

Comment:
so rather than making these separate functions exposed to python, I think the proper "torch" thing to do is to check (in the torch C++ layer) if dof gradients are required, then dispatch to the appropriate C++ function.

Reply:
ooooh, hmm